crates/oxc_linter/src/rules/react/jsx_filename_extension.rs

```diff
@@ -0,0 +1,288 @@
+    /// The set of allowed extensions is configurable. By default `'.jsx'` is allowed. If you wanted to allow both `'.jsx'` and `'.tsx'`, the configuration would be:
+    /// ```js
+    /// "rules": {
+    ///     "react/jsx-filename-extension": ["error", { "extensions": [".jsx", ".tsx"] }]
+    /// }
+    /// ```
+    ///
+    /// #### `ignoreFilesWithoutCode` (default: `false`)
+    /// If enabled, files that do not contain code (i.e. are empty, contain only whitespaces or comments) will not be rejected.
+    /// ```js
+    /// "rules": {
+    ///     "react/jsx-filename-extension": ["error", { "ignoreFilesWithoutCode": true }]
+    /// }
+    /// ```
+    JsxFilenameExtension,
+    react,
+    restriction,
+    pending
+);
+
+impl Rule for JsxFilenameExtension {
+    fn from_configuration(value: Value) -> Self {
+        let config = value.get(0);
+
+        let ignore_files_without_code = config
+            .and_then(|config| config.get("ignoreFilesWithoutCode"))
+            .and_then(Value::as_bool)
+            .unwrap_or(false);
+
+        let allow = config
+            .and_then(|config| config.get("allow"))
+            .and_then(Value::as_str)
+            .map(AllowType::from)
+            .unwrap_or_default();
+
+        let extensions = config
+            .and_then(|v| v.get("extensions"))
+            .and_then(Value::as_array)
+            .map(|v| {
+                v.iter()
+                    .filter_map(serde_json::Value::as_str)
+                    .filter(|&s| s.starts_with('.'))
+                    .map(|s| &s[1..])
+                    .map(CompactStr::from)
+                    .collect()
+            })
+            .unwrap_or(vec![CompactStr::from("jsx")]);
+
+        Self(Box::new(JsxFilenameExtensionConfig { allow, extensions, ignore_files_without_code }))
+    }
+
+    fn run_once(&self, ctx: &LintContext) {
+        let jsx_elt = ctx
+            .nodes()
+            .iter()
+            .find(|&&x| matches!(x.kind(), AstKind::JSXElement(_) | AstKind::JSXFragment(_)));
+        let file_extension = ctx.file_path().extension().and_then(OsStr::to_str).unwrap_or("");
+        let has_ext_allowed = self.extensions.contains(&CompactStr::new(file_extension));
+
+        if jsx_elt.is_some() {
+            if !has_ext_allowed {
+                let span_elt = jsx_elt.map(GetSpan::span);
+                ctx.diagnostic(no_jsx_with_filename_extension_diagnostic(
+                    file_extension,
+                    span_elt.unwrap(),
+                ));
+            }
+            return;
+        }
+
+        if matches!(self.allow, AllowType::AsNeeded) && has_ext_allowed {
+            let Some(root) = ctx.nodes().root_node() else {
+                return;
+            };
+            let AstKind::Program(program) = root.kind() else { unreachable!() };
+            if self.ignore_files_without_code && program.body.is_empty() {
+                return;
+            }
```

Comment:
We want to avoid using `unreachable!()`, because it could cause a panic at runtime. instead of checking the program node directly we could check `ctx.nodes().len()` and see if it is greater than 1, indicating it is not empty.

Reply:
thank you, i fix it.